### PR TITLE
fix(datepicker): gray out filtered years in multi-year view

### LINKED ISSUE #9516: MatDatePicker years not blacked out
#### Feature request
(Updated with issue template.)

#### Expected behavior?
I'm pretty sure this one should be easily replicable: when using the MatDatePicker and max/min or matDatepickerFilter, any months without available dates are disabled in the month chooser, but years are all enabled in the year chooser. 

#### What is the current behavior?
This is demonstrated in the docs in the 'Datepicker with filter validation' section here: 
https://material.angular.io/components/datepicker/overview#date-validation
ie. https://stackblitz.com/angular/eyqovodandr
- drill up to years and <2000 and >2021 are enabled.

#### What are the steps to reproduce?
Open link above, open datepicker popup, drill up to years.

#### What is the use-case or motivation for changing an existing behavior?
Useful to browse available years, as we have with available months.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
All

#### Is there anything else we should know?

Bonus points (now linked in below comment): has anyone thought about adding a range feature into MatDatePicker - ie. start & end dates within a single datepicker control?

### PR DESCRIPTION
**Pull request by julianobrasil**
Closes #9516 

image

### CODE REVIEW

File: src/lib/datepicker/multi-year-view.spec.ts [hunk 1]
```diff
@@ -71,6 +72,27 @@ describe('MatMultiYearView', () => {
...
+    beforeEach(() => {
+      fixture = TestBed.createComponent(MultiYearViewWithDateFilter);
+      fixture.detectChanges();
+
+      let multiYearViewDebugElement = fixture.debugElement.query(By.directive(MatMultiYearView));
+      multiYearViewNativeElement = multiYearViewDebugElement.nativeElement;
+      testComponent = fixture.componentInstance;
+    });
+
+    it('should disabled years with no enabled days', () => {
```
Comment (donroyco) [inline src/lib/datepicker/multi-year-view.spec.ts]: *disable

File: src/lib/datepicker/multi-year-view.ts [hunk 2]
```diff
@@ -17,9 +17,9 @@ import {
   Output,
   ViewEncapsulation
 } from '@angular/core';
-import {DateAdapter} from '@angular/material/core';
-import {MatCalendarCell} from './calendar-body';
-import {createMissingDateImplError} from './datepicker-errors';
+import { DateAdapter } from '@angular/material/core';
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: undo

File: src/lib/datepicker/multi-year-view.ts [hunk 3]
```diff
@@ -47,9 +47,9 @@ export class MatMultiYearView<D> implements AfterContentInit {
   set activeDate(value: D) {
     let oldActiveDate = this._activeDate;
     this._activeDate =
-        this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
+      this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: undo

File: src/lib/datepicker/multi-year-view.ts [hunk 4]
```diff
@@ -79,8 +79,8 @@ export class MatMultiYearView<D> implements AfterContentInit {
   /** The year of the selected date. Null if the selected date is null. */
   _selectedYear: number | null;
 
-  constructor(@Optional() public _dateAdapter: DateAdapter<D>,
-              private _changeDetectorRef: ChangeDetectorRef) {
+  constructor( @Optional() public _dateAdapter: DateAdapter<D>,
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: undo

File: src/lib/datepicker/multi-year-view.ts [hunk 5]
```diff
@@ -112,9 +112,9 @@ export class MatMultiYearView<D> implements AfterContentInit {
   _yearSelected(year: number) {
     let month = this._dateAdapter.getMonth(this.activeDate);
     let daysInMonth =
-        this._dateAdapter.getNumDaysInMonth(this._dateAdapter.createDate(year, month, 1));
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: undo

File: src/lib/datepicker/multi-year-view.ts [hunk 6]
```diff
@@ -124,7 +124,26 @@ export class MatMultiYearView<D> implements AfterContentInit {
...
+    return new MatCalendarCell(year, yearName, yearName, this._isYearEnabled(year));
+  }
+
+  /** Whether the given year is enabled. */
+  private _isYearEnabled(year: number) {
+    if (!this.dateFilter) {
+      return true;
+    }
+    
+    let firstOfYear = this._dateAdapter.createDate(year, 0, 1);
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: const

**Comment by julianobrasil** [comment]
Comments addressed. "SHIFT+ALT+F" from habit... :blush:

File: src/lib/datepicker/multi-year-view.ts [hunk 7]
```diff
@@ -49,7 +49,7 @@ export class MatMultiYearView<D> implements AfterContentInit {
     this._activeDate =
         this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
     if (Math.floor(this._dateAdapter.getYear(oldActiveDate) / yearsPerPage) !=
-        Math.floor(this._dateAdapter.getYear(this._activeDate) / yearsPerPage)) {
+      Math.floor(this._dateAdapter.getYear(this._activeDate) / yearsPerPage)) {
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: missed this one

File: src/lib/datepicker/multi-year-view.ts [hunk 8]
```diff
@@ -114,7 +114,7 @@ export class MatMultiYearView<D> implements AfterContentInit {
     let daysInMonth =
         this._dateAdapter.getNumDaysInMonth(this._dateAdapter.createDate(year, month, 1));
     this.selectedChange.emit(this._dateAdapter.createDate(year, month,
-        Math.min(this._dateAdapter.getDate(this.activeDate), daysInMonth)));
+      Math.min(this._dateAdapter.getDate(this.activeDate), daysInMonth)));
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: missed this one

File: src/lib/datepicker/multi-year-view.spec.ts [hunk 9]
```diff
@@ -71,6 +72,27 @@ describe('MatMultiYearView', () => {
...
+  describe('multi year view with date filter', () => {
+    let fixture: ComponentFixture<MultiYearViewWithDateFilter>;
+    let testComponent: MultiYearViewWithDateFilter;
+    let multiYearViewNativeElement: Element;
+
+    beforeEach(() => {
+      fixture = TestBed.createComponent(MultiYearViewWithDateFilter);
+      fixture.detectChanges();
+
+      let multiYearViewDebugElement = fixture.debugElement.query(By.directive(MatMultiYearView));
```
Comment (rafaelss95) [inline src/lib/datepicker/multi-year-view.spec.ts]: `const`

File: src/lib/datepicker/multi-year-view.spec.ts [hunk 10]
```diff
@@ -71,6 +72,27 @@ describe('MatMultiYearView', () => {
...
+      multiYearViewNativeElement = multiYearViewDebugElement.nativeElement;
+      testComponent = fixture.componentInstance;
+    });
+
+    it('should disablex years with no enabled days', () => {
+      let cells = multiYearViewNativeElement.querySelectorAll('.mat-calendar-body-cell');
```
Comment (rafaelss95) [inline src/lib/datepicker/multi-year-view.spec.ts]: `const`

File: src/lib/datepicker/multi-year-view.spec.ts [hunk 11]
```diff
@@ -84,3 +106,16 @@ class StandardMultiYearView {
...
   @ViewChild(MatYearView) yearView: MatYearView<Date>;
 }
+
+@Component({
+  template: `<mat-multi-year-view [activeDate]="activeDate" [dateFilter]="dateFilter"></mat-multi-year-view>`
+})
+class MultiYearViewWithDateFilter {
+  activeDate = new Date(2017, JAN, 1);
+  dateFilter(date: Date) {
+    if (date.getFullYear() == 2017) {
```
Comment (rafaelss95) [inline src/lib/datepicker/multi-year-view.spec.ts]: What about `return date.getFullYear() !== 2017;`

File: src/demo-app/datepicker/datepicker-demo.ts [hunk 12]
```diff
@@ -31,7 +31,7 @@ export class DatepickerDemo {
   lastDateInput: Date | null;
   lastDateChange: Date | null;
 
-  dateFilter = (date: Date) => date.getMonth() % 2 == 1 && date.getDate() % 2 == 0;
+  dateFilter = (date: Date) => date.getFullYear() % 2 == 0 && date.getMonth() % 2 == 1 && date.getDate() % 2 == 0;
```
Comment (rafaelss95) [inline src/demo-app/datepicker/datepicker-demo.ts]: Could be simplified to `!date.getFullYear() && date.getMonth() % 2 == 1 && !date.getDate() % 2;`

Comment (mmalerba) [reply src/demo-app/datepicker/datepicker-demo.ts]: Assume you meant: `!date.getFullYear() % 2 && date.getMonth() % 2 && !date.getDate() % 2;`?

Comment (rafaelss95) [reply src/demo-app/datepicker/datepicker-demo.ts]: Ups, yes. Bad copy/paste :D

**Pull request merged**